### ISSUE TITLE
Seems to not work well in slim areas

### ISSUE THREAD

**Issue by Sticcolet** [hako-mikan/sd-webui-regional-prompter#57]
I find it may not work correctly when you divide the area into slim piece. For example, while using 1:10:1 to segment a square, It will  generate some strange results.

**Comment by hako-mikan** [comment]
This may be a problem caused by Stable Diffusion's image generation process. If you are generating a 1:10:1 ratio, it is better to set the length of side to be divided to a multiple of 12*8.

**Comment by Sticcolet** [comment]
Good idea, I will try later. By the way, is there any way to make SD generate elements in different corner? For example, I divide a picture into 1;6,1,2;1,2, and I want to generate some scenery in piece 2 and 4 but left 0,1,3 empty. I have tried several times but failed.
![0](https://user-images.githubusercontent.com/33365438/232471204-b0854e5c-8fe9-40a2-9a31-41b49822cf3b.png)

**Comment by hako-mikan** [comment]
Its difficult to separate for attention mode.
Latent mode is more rigorous in area division, so you may want to try that.

**Comment by Sticcolet** [comment]
Ok, I will take a try, thanks a lot!

**Comment by Symbiomatrix** [comment]
> By the way, is there any way to make SD generate elements in different corner? For example, I divide a picture into 1;6,1,2;1,2, and I want to generate some scenery in piece 2 and 4 but left 0,1,3 empty. I have tried several times but failed.

You mean completely empty or put something else?
Have you tried using the same structure in the negative prompt and negating whichever scenery you're using for 0/1/3?

**Comment by Sticcolet** [comment]
> > By the way, is there any way to make SD generate elements in different corner? For example, I divide a picture into 1;6,1,2;1,2, and I want to generate some scenery in piece 2 and 4 but left 0,1,3 empty. I have tried several times but failed.
> 
> You mean completely empty or put something else? Have you tried using the same structure in the negative prompt and negating whichever scenery you're using for 0/1/3? Meaning, prompt="scenery ADDROW character" neg=", ADDROW (scenery)"

In fact I prefer some solid color backgrounds like white and grey, but I know its difficult because of SDs generation process, so I used some empty background like sky or wall instead. Sadly, at present, it seems that the effect is still not good.

**Comment by Sticcolet** [comment]
Well, sometimes it works. But now there is another question, it may generation irrelevant things.
For examlpe, when using 1,2,1;2,2,1 to dividing and using promote="masterpiece, no human, ADDCOMM, mountain, sky, simple background, ADDCOL, sunshine, cloud, blank,simple background,ADDROW, small butterfly on branch, plum blossoms, ADDCOL, blank,simple background,", its aim to generate flower at 2 and left 0,1 and 3 empty. But the result always contain a cat  or some other strange things.
![grid-0189-1027284650-masterpiece, no human, ADDCOMM,_ mountain, sky, simple background, ADDCOL,_ sunshine, cloud, blank,simple background,ADDROW,_ sm](https://user-images.githubusercontent.com/33365438/232706701-31a01a38-944f-47c6-90cb-c3c7e1c9e597.png)
![0](https://user-images.githubusercontent.com/33365438/232706734-04e69ed6-3bc1-414d-b9f7-8453ef098d8b.png)

**Comment by Symbiomatrix** [comment]
I'm not quite sure "no human" will achieve the intended result due to the tokenisation process. Adding character / subject / animal / human to the negative seems better. Likewise, it seems to be making "simple" shapes on a generic background. I'm pretty bad at prompting myself, but that was my impression in general.

**Comment by Sticcolet** [comment]
> I'm not quite sure "no human" will achieve the intended result due to the tokenisation process. Adding character / subject / animal / human to the negative might work better. Likewise, it seems to be making "simple" shapes on a generic background. I'm pretty bad at prompting myself, but that was my impression in general.

I have tried add "human", "animal" or "element" in negtive like "no human, bad, error, low quality, 1girl, ADDCOMM, ADDCOL,
 ADDROW, animal, human, ADDCOL," but seems useless. Maybe thats cause by the generation process of SD. Empty area will cause larger loss so that SD fell uncomfortable. (ovo)
By the way, latent mode cant work correctly in this condition. It seem generate 4 areas independently. My batch size is 4 and batch count is also 4.
![grid-0192-428179514-masterpiece, no human, ADDCOMM,_ mountain, sky, simple background, ADDCOL,_ sunshine, cloud, blank,simple background,ADDROW,_ sm](https://user-images.githubusercontent.com/33365438/232723093-180156e4-1d0f-42f5-99fb-611cf196dc6a.png)

**Comment by Sticcolet** [comment]
Now I found a more strange thing, I accidentally used mismatched prompts and segmentation, but it actually produced better results. My promote is “masterpiece, <lora:animeLineartMangaLike_v30MangaLike:0.5>, (no humans:1.5),  <lora:standingFullBodyWithBackgroundStyle_v10Offset:0.5>ADDCOMM, sky, simple background, ADDCOL, sunshine, cloud, blank,simple background,ADDROW, small butterfly on branch, plum blossoms, ADDCOL, blank,simple background,nothing,” for a 2*2 area. But my segmentation is 1,3,3,2;1,3,3,2;1,3,3,2 for 3*3 area. BUT！The results seems better!
![Uploading 1.png…]()

**Comment by Symbiomatrix** [comment]
I can't answer reliably about latent mode as I'm still trying to figure out some (most) things on how it works.
I can say that the region code has a safeguard for mismatching prompt and segmentation: if you use ADDCOL / ADDROW syntax, then it will take that to be the base region, and try to fill out the values from segmentation - for example, the first (1,3,3,2) will be used for row 1 and at most 3 columns, then either discard the excess values or copy the last one to missing values (broadcast).
There's a line which shows up in debug mode, "ratios : {aratios}" which prints what ratios it ended up using. But I might've forgotten to add a printing function. I'll let you know.

**Comment by Sticcolet** [comment]
I reupload the result picture.
So if the prompts dont match the segmentation, it will prioritize the segmentation method described by the prompts and ignore other segmented parts? I will take a try.

**Comment by Symbiomatrix** [comment]
Hmm... That's an interesting observation indeed. Could be a feasible explanation for the problem - the lora weight is compounded due to being applied per region, for the whole thing.
It's a difficult to see how exactly the model is built from the debug steps (need something like keras' .summary), but I could atleast try to see if the input weights increase with the repeated layers, implying a chaining of some kind, or if they're independent.

**Comment by Sticcolet** [comment]
> Hmm... That's an interesting observation indeed. Could be a feasible explanation for the problem - the lora weight is compounded due to being applied per region, for the whole thing. It's difficult to see how exactly the model is built from the debug steps (need something like keras' .summary), but I could atleast try to see if the input weights increase with the repeated layers, implying a chaining of some kind, or if they're independent (applied in channels).

Looking forward to your good news! And is it possible to dividing regions with irregular shapes? For example, split a square to a smaller square and a "L"? I guess its possible to make SD focus its attention to a corner.

**Comment by Symbiomatrix** [comment]
Directly, no. It's not currently intended to be on par with inpainting. But repeating the prompt should achieve similar results in this case (save for loras obviously), ex
[H] 1 ADDROW 1 ADDCOL 2

**Comment by Sticcolet** [comment]
All right, that what I do now. I may try it by myself.

**Comment by Symbiomatrix** [comment]
Theoretically, I guess it could be done without extreme effort in the backend: Makefilters is already a mask in latent, build a similar mask method in attention instead of torchcatenating subregions as we currently do in attention. And a polygon lasso selection in a canvas to make custom numbered subregions would be pretty cool.
Unfortunately, I have only very basic knowledge of frontend in general, and no clue how gradio could support such a thing, if it's even possible.

**Comment by Symbiomatrix** [comment]
@Sticcolet So I've written this small prototype which allows creating polygon regions based on outlines (though fully colouring works too), and automatically assigns colours to each which can later be used to mask the regions precisely.
What do you think? How does this method compare with inpainting?

https://github.com/Symbiomatrix/gradio-test-polygon

**Comment by Sticcolet** [comment]
Sorry for the late reply, I will try this method. Thank you for your work!

**Comment by Symbiomatrix** [comment]
Sorry, I'm not entirely familiar with these terms. What do you mean? What sort of problems have you seen using irregular shapes in inpainting exactly?

**Comment by Sticcolet** [comment]
> Sorry, I'm not entirely familiar with these terms. What do you mean? What sort of problems have you seen using irregular shapes in inpainting exactly?

Actually, you don't have to worry about it. These are just some special situations I encountered during my own use, and I think it may be related to the proportion of the mask area in the entire image.

For example, for a mask like the one below, I would like SD
![down](https://user-images.githubusercontent.com/33365438/233977090-1d049873-fa09-4495-b745-73e95fe13259.gif)
 to paint in a white area, but there is approximately a half chance that the painted area will turn into a red triangular area.

**Comment by Symbiomatrix** [comment]
Good news everyone, I found a workaround for the sketch refresh bug.
Got a working alpha version (sans latent, it's difficult to test on my version since I have to restart the whole webui every time), I'll upload it to a side branch.
